### DISCUSSION TITLE
Progress report?

### DISCUSSION THREAD

**Discussion by audiamus** [Mbucari/AAXClean.Codecs#2] [Q&A] [answered]
Quick question on silence detection: I only had a brief look at it so far, but could not detect an option for progress callbacks, like AAXClean itself has. Is it hidden or can it be added?

**Comment by Mbucari** [comment, marked as answer]
I think that would be pretty easy to add into the existing code. I tested and optimized the hell out of that silence detection routine so it runs just about as fast as the AAC can be decoded, but it runs on its own thread. Just be aware that any UI stuff done with that callback will be on a different thread than the invoker.

**Comment by Mbucari** [comment]
I added an overload that accepts a callback in version 0.1.7.

**Reply by audiamus** [reply]
Very good. I hope I'll find the time to write a short test in the next few days. Different threads won't be a problem. My software is multi-threaded throughout. With my current FFmpeg silence detection I run parallel sub-processes and split the source file for that. Not only for performance reasons but for a shortcoming in the FFmpeg silence output. This is text with a fixed field width, and for an unsplit longer book, the fractional part gets fewer digits with increasing time values, and results become inaccurate. With TimeSpan, no such problems are expected, of course.

**Reply by Mbucari** [reply]
My silence detection as it is currently written cannot be parallelized. I can probably rewrite it to work in parallel, but the difficulty with that is that AAC audio frames require the previous decoded frames as a seed to properly decode the current frame. I know skipping around as possible, but that is something I would have to figure out.

Currently, I can achieve about a 1,800 x speed up with the silence detect routine on a 4th generation Intel i7.

**Reply by audiamus** [reply]
No reason for parallelization on your side.  If your new algo is really fast, I won't need it. Otherwise I would split the input file first and run silence detection in parallel for the individual parts, like I do with FFmpeg.

**Comment by audiamus** [comment]
I managed to write a short test: And it's looking very good!

Detected silences appear to be exactly the same as with FFmpeg (-30db, 025s), plus the increased accuracy as a bonus. FFmpeg is rounding down to 10 ms after 1000 s. And it's faster than FFmpeg. Total silence detection time is still longer with AAXClean, but only because of my parallelization in AAX Audio Converter.

Also tested the progress callback, very handy!
